spec: ipfs:// and ipns:// URI schemes (#544)
ipfs/specs: src/ipfs-uri.md

  any request, or reject the URI; resolving the non-canonical form in place would split one
  content root into several origins.
- The content root SHOULD NOT exceed **63 characters**. The limit is DNS-specific: a single DNS
  label caps at 63 characters, and staying within it is what lets the root serve as a
  Subdomain Gateway (:cite[subdomain-gateway]) label and its
  origin. A local resolver or a native `ipfs://` implementation is not bound by DNS and MAY
  support a longer root, for example by mapping it to a localhost subdomain that public DNS
  never sees. Such roots cripple interoperability, though, and SHOULD be kept to internal use.
  An implementation that cannot rely on such a mapping SHOULD reject a root longer than 63
  characters.
- The content root MUST NOT contain any code point that the WHATWG URL Standard lists as a
  [forbidden host code point](https://url.spec.whatwg.org/#forbidden-host-code-point) or
  [forbidden domain code point](https://url.spec.whatwg.org/#forbidden-domain-code-point)
  (notably `%`).
- A producer MUST NOT emit a case-sensitive encoding: a **CIDv0** (`base58btc` `Qm...`) or a

>   `base58btc` CIDv1 (`z...`) is corrupted once the authority is case-folded (see

Reviewer: ```suggestion
  `base58btc` CIDv1 (`z...`) gets corrupted when the authority is case-folded (see
```

Author: Keeping "is corrupted once": "once" marks case-folding as the point of no return, which is what separates this bullet from the next one about input caught while its case is intact. Shout if it reads stilted to you.

  concatenation of the authority and the path segments. The URI path MUST NOT repeat the `/ipfs/`
  namespace prefix, which the scheme already implies; a path segment that happens to be named
  `ipfs` further down the DAG is ordinary content and is left alone. An empty path and a path of
  `/` both address the content root itself. The path is passed as-is: this specification adds no
  encoding or decoding beyond what the WHATWG URL parser already does. What the segments mean is
  decided by the content root, as described below.
- **Query.** A query, if present, is carried through to the resolver unchanged, with no
  encoding or decoding beyond what the WHATWG URL parser already does.
- **Fragment.** A fragment is a client-side component of the URL and is passed through: when the
  URI is mapped to a gateway URL, the fragment is carried onto that URL unchanged. It MUST NOT be
  sent in retrieval requests, and it plays no part in resolution.

### How the content root drives traversal

A CID is self-describing, and its multicodec is what tells a resolver both how to decode the block
the CID addresses and how to walk `{path}` out of that block. Nothing in the URI selects this

> behavior, and nothing needs to: the codec in the content root does.

Reviewer: ```suggestion
behavior, and nothing needs to: the codec prefix within the CID of the content root is all the signaling needed.
```

Author: Keeping the shorter form: the CID *is* the authority, so putting the signal "within the CID" puts it back inside the URI, which is what the first half of the sentence rules out. "Prefix" is also loose, in a CIDv1 the multicodec is the second varint after the version, and CIDv0 has no codec field at all.

- `dag-pb` is read as UnixFS (:cite[unixfs]), so path segments are directory and file names.
- `dag-cbor` holds structured data, and a path segment names a field within it. A field whose value
  is a CID carried under
  [CBOR tag 42](https://datatracker.ietf.org/doc/draft-caballero-cbor-cbor42/) is a link, and
  following it moves the traversal into a different document.

> - `raw` is opaque bytes with no links to follow, so a non-empty path has nowhere to go.

Reviewer: ```suggestion
- `raw` is opaque bytes with no links to follow, so a non-empty path has nowhere to go.
- no other codecs can be taken as signals to path traversal metadata, so for any other codec values, non-empty paths will be similarly unresolvable.
```
Just to make this more explicit (wording not important, feel free to reword)

Author: Worth stating, but "no other codecs" isn't true: `dag-json` is traversable too, links are `{"/": "<cid>"}`.

Added the general rule after the list in ea07e7b instead: traversal goes only as far as the codec defines links, so a codec with no links, or one the resolver doesn't recognize, leaves a non-empty path unresolvable.

This applies at every step, not only the first. Each link the traversal follows carries its own CID
with its own codec, so the rules can change partway down a path: a `dag-cbor` root can link to a

> `dag-pb` node, after which the remaining segments are UnixFS names. Path Gateway
> (:cite[path-gateway]) defines the traversal itself, including how an unresolvable path is reported.

Reviewer: ```suggestion
`dag-pb` node, after which the remaining segments are UnixFS names. The Path Gateway
(:cite[path-gateway]) specification defines in detail the traversal itself, 
including how an unresolvable path is reported.
```

Author: Keeping this one as-is. `:cite` renders as a bracketed reference, so "The Path Gateway ([path-gateway]) specification" splits the name from the noun. The bare form also matches how Subdomain Gateway and Trustless Gateway are cited elsewhere in the file.

carried through; the fragment stays client-side.

A resolver that retrieves blocks itself, over a Trustless Gateway (:cite[trustless-gateway]) or a
peer-to-peer transport, MUST verify them: hash every block it receives and walk the DAG from the
content root, so that each block on the path is accounted for. Verification is what lets an
`ipfs://` URI be trusted regardless of where the bytes came from.

Retrieval through a trusted Path or Subdomain Gateway is the exception: the gateway performs that
verification and the client takes its word for it. An implementation that resolves `ipfs://` this
way SHOULD make the gateway it trusts visible to the user, because the immutability guarantee then
rests on that gateway rather than on the URI.

A native `ipfs://` implementation SHOULD resolve content paths with the same semantics as a
Path Gateway, so that the same URI resolves consistently whether it is opened by a native handler
or handed to a gateway. In a browser-like context, a native implementation SHOULD also mirror the
origin isolation that a Subdomain Gateway provides, giving each content root its own origin (see

> [Origin](#origin)).

Reviewer: I think I know what this means, but I'm a little confused by what that SHOULD would look like in practice.  Maybe a parenthetical and non-normative example like `.  (For example, a browser engine/SPA would treat each root CID as a distinct secure context by.... .)` would help?

Author: Fair, the sentence didn't say how. Reworded in ea07e7b to name the mechanism: the origin comes from the scheme and the content root alone, and the Origin section spells out the two serializations that work.

Stayed off "secure context" though, this doc uses that for why `*.localhost` needs no TLS cert, which is a different property.

## Notes for implementers

This section is non-normative: it explains the reasoning behind the rules above and adds no
new requirements.

### Reading `ipfs://` as a WHATWG URL

The `//` is what makes the content root an authority. Neither `ipfs` nor `ipns` is a
[special scheme](https://url.spec.whatwg.org/#special-scheme), and the two forms parse very

> differently:

Reviewer: ```suggestion
The `//` is what makes the content root an authority. Neither `ipfs` nor [`ipns`](./ipns-uri.md)
 is a [special scheme](https://url.spec.whatwg.org/#special-scheme), and the two forms 
 (with and without the `//`) parse very differently by design:
```

Author: Added "(with and without the `//`)" in ea07e7b.

Skipped the `ipns` link: the generator leaves relative `.md` links as-is, so `./ipns-uri.md` would 404 at `/ipfs-uri/`. The doc links `ipns://` to https://specs.ipfs.tech/ipns-uri/ higher up.
